### ISSUE TITLE
`FileArchive` state does not survive `refresh` operations

### ISSUE THREAD

**Issue by Navigatron** [pulumi/pulumi#15862]
### What happened?

I am uploading a local folder to Azure blob storage as a zip file. The `source` property on an `azure-native:storage:Blob` resource accepts an `asset`, which I have set to a `FileArchive` object.

1. The `up` operation runs just fine
2. A subsequent `refresh` also runs just fine, but mutates the `source` input (specifically, it removes the `path` property)
3. Any following `preview` or `up` throws: `Invalid archive encountered when unmarshaling resource property`
    - [In /runtime/rpc.ts](https://github.com/pulumi/pulumi/blob/master/sdk/nodejs/runtime/rpc.ts#L619)

Statefile diff before and after refresh:
![image](https://github.com/pulumi/pulumi/assets/6775862/02143646-2480-441b-9000-ee9eb9b2aebd)

Explosion:
![image](https://github.com/pulumi/pulumi/assets/6775862/d9c98cb6-2ba0-408d-9986-5259cc9050ed)

### Example

This should be 95% of the way to a working POC:

```ts
import {fileURLToPath, URL} from 'node:url';
import {LocalWorkspace} from '@pulumi/pulumi/automation/index.js';
import {ResourceGroup} from '@pulumi/azure-native/resources/resourceGroup.js';
import {StorageAccount} from '@pulumi/azure-native/storage/storageAccount.js';
import {BlobContainer} from '@pulumi/azure-native/storage/blobContainer.js';
import {FileArchive} from '@pulumi/pulumi/asset/archive.js';
import {Blob} from '@pulumi/azure-native/storage/blob.js';

async function build(){
	// Resource Group
	const resourceGroup = new ResourceGroup('filearchivepoc', {
		resourceGroupName: 'filearchivepoc',
		location: 'eastus2',
	});

	// Storage Account
	const storageAccount = new StorageAccount('fileArchivePoc', {
		kind: 'StorageV2',
		sku: {
			name: 'Standard_LRS',
		},
		resourceGroupName: resourceGroup.name,
	});

	// Blob Container
	const blobContainer = new BlobContainer('zips', {
		containerName: 'zips',
		resourceGroupName: resourceGroup.name,
		accountName: storageAccount.name,
	});

	// Create the FileArchive - will need a folder called "function"
	const functionFolder = fileURLToPath(new URL('./function', import.meta.url));
	const asset = new FileArchive(functionFolder);

	// Blob
	const blob = new Blob('function.zip', {
		blobName: 'function.zip',
		resourceGroupName: resourceGroup.name,
		accountName: storageAccount.name,
		containerName: blobContainer.name,
		source: asset,
	});

	return blob;
}

async function main(){
	const pulumiArgs = {
		stackName: 'FileArchiveDebug',
		projectName: 'FileArchiveDebug',
		program: build,
	};
	const pulumiStack = await LocalWorkspace.createOrSelectStack(pulumiArgs);

	// Deploy
	await pulumiStack.up();

	// Refresh; corrupting state
	await pulumiStack.refresh();

	// Explode
	await pulumiStack.up();
}

main();

```

### Output of `pulumi about`

We are using the pulumi automation API.

- @pulumi/pulumi:3.89.0
- @pulumi/azure-native:1.104.0 (this is a long story; only used for vm scalesets)
- @pulumi/azure-native:2.4.0

Output of pulumi about:
```
CLI          
Version      3.74.0
Go Version   go1.20.5
Go Compiler  gc

Host     
OS       ubuntu
Version  20.04
Arch     x86_64

Backend        
Name           <internal project name>
URL            azblob://<internal project name>
User           azureuser
Organizations  

Pulumi locates its logs in /tmp by default
warning: Failed to read project: no Pulumi.yaml project file found (searching upwards from /home/azureuser/agent/_work/2/s/environments/dev1). If you have not created a project yet, use `pul
umi new` to do so: no project file found
warning: Failed to get information about the current stack: no Pulumi.yaml project file found (searching upwards from /home/azureuser/agent/_work/2/s/environments/dev1). If you have not crea
ted a project yet, use `pulumi new` to do so: no project file found
```

### Additional context

I've found a few other github issues that may be relevant / have some clues:
- https://github.com/pulumi/pulumi/issues/6691
- https://github.com/pulumi/pulumi-azure-native/issues/642
- https://github.com/pulumi/pulumi/issues/3018

I'll see if I can add zipping the folder to that sub-project's build step; then we may be able to use just `FileAsset` instead; and we'll see if that works.

### Contributing

Vote on this issue by adding a 👍 reaction. 
To contribute a fix for this issue, leave a comment (and link to your pull request, if you've opened one already).

**Comment by Navigatron** [comment]
It looks like this should have been fixed in https://github.com/pulumi/pulumi/commit/142a25216b3043bc90a4cd83fe2c3d2578c628de , which would be included in releases 3.87 and up. Our code uses 3.89, but it looks like the build vm has 3.74. This shouldn't invalidate anything, but just to triple check, I'll update everything and re-test.

**Comment by Navigatron** [comment]
Lo and behold, I updated, and everything is working perfectly. I was so pessimistic, and this has just made my day.

@Zaid-Ajaj , you are my hero. Thank you so much. Your fix is working exactly as intended, and I am super super happy about it. If you ever find yourself in Rochester NY, I owe you many beers (or rounds of golf, you pick!) <3
